### ISSUE TITLE
Sharp is not working with "Bun build"

### ISSUE THREAD

**Issue by buzzy** [lovell/sharp#4283]
When compiling a binary using "Bun build", it seems that the resulting binary does not contain the "binary" of Sharp. It still tries to load it from ./node_modules. If I copy the node_modules directory together with the resulting binary from "Bun build", it runs without issues. I am not sure if this is an issue with "Bun build" or with Sharp, but other node-related binaries from node_modules are working for me.

**Comment by lovell** [comment]
It's very unlikely that `bun build --compile` will work with sharp as it attempts to create a single executable, however sharp and libvips are provided as shared dynamic libraries.

**Comment by buzzy** [comment]
> It's very unlikely that `bun build --compile` will work with sharp as it attempts to create a single executable, however sharp and libvips are provided as shared dynamic libraries.

Well, so does sqlite, but that works with Bun. I am not talking about the built in version, but the node module. Bun embeds the lib into the binary on --compile

**Comment by lovell** [comment]
Bun provides its own sqlite implementation that does not depend on 3rd party shared libraries so it's not quite the same thing.

My previous advice to people attempting to do something similar with `deno compile` is to experiment with WebAssembly - please see https://github.com/lovell/sharp/issues/3912#issuecomment-1866214392

**Comment by buzzy** [comment]
> Bun provides its own sqlite implementation that does not depend on 3rd party shared libraries so it's not quite the same thing.
> 
> My previous advice to people attempting to do something similar with `deno compile` is to experiment with WebAssembly - please see https://github.com/lovell/sharp/issues/3912#issuecomment-1866214392

Like I wrote in the earlier post; I am not talking about the built-in sqlite library. I am talking about the node-lib of sqlite that also works with Bun. Any case, I now understand the issue. I have opened an issue with the Bun team so that they can include the dynamic lib inside the binary. Thanks!

**Comment by LetrixZ** [comment]
> My previous advice to people attempting to do something similar with `deno compile` is to experiment with WebAssembly - please see [#3912 (comment)](https://github.com/lovell/sharp/issues/3912#issuecomment-1866214392)

I patched `sharp.js` to force the Wasm library to load but, upon invocation, my program froze.
```js
'use strict';
module.exports = require('@img/sharp-wasm32/sharp.node')
```

Do you know how should I debug this?

Thanks.

**Comment by infrahead** [comment]
I got this working as a sort of experiment previously and can probably be done even easier now, with very recent improvements in https://bun.sh/docs/bundler/executables#embed-assets-files.   As the sharp .node addon builds fine, the snag is in the libvips dll/so library.  After @lovell answered https://github.com/lovell/sharp/issues/4023 it was clear the path for now was doing some imports/scripting to make those files embedded in the binary, and then update the reference from sharp.

**Comment by LetrixZ** [comment]
After experimenting I managed to get it working with this:

I use this script to make a file which basically crates a TS file containing an import line with `{ type: 'file' }`.

```ts
# embed-sharp.ts
import { runtimePlatformArch } from './node_modules/sharp/lib/libvips';

const runtimePlatform = runtimePlatformArch();

const path = `./node_modules/sharp/src/build/Release/sharp-${runtimePlatform}.node`;

// eslint-disable-next-line @typescript-eslint/no-require-imports
require(path);
Bun.write('./compile/sharp.ts', `import '${path.replace('./', '../')}' with { type: 'file' }`);
```

I then import the generated file in my entrypoint:

```ts
...
import './compile/sharp';

...
```

This is my build script:

```sh
# build.sh
bun ./embed-sharp.ts &&
bun build --outfile my_binary --minify --sourcemap --compile ./entrypoint.ts
```

Importing the package `@img/sharp-${runtimePlatform}/sharp.node` didn't work as then the lib files wouldn't be included.

I only tried on macOS. Need to test if this works for other OSes.

**Comment by LetrixZ** [comment]
This might not work. It complains that `libvips-cpp.so.42` is missing when trying on a fresh system.

**Comment by ev3nst** [comment]
i dont think this is exclusively related to bun. i tried with node js pkg and same/similar error messages.

**Comment by ClarenceL** [comment]
So, no go with bun for now I guess?

**Comment by jo-chemla** [comment]
Sorry to revive a closed thread, potential fix described here https://github.com/lovell/sharp/issues/4377#issuecomment-2819585265


> Just a quick update. Found a solution. I was missing `--config package.json` in the `npx` command:
> 
> ```
> npx pkg index.js --config package.json --targets node20-macos-arm64 --output index
> ```
> 
> And added assets to `package.json`:
> 
> ```
>  "pkg": {
>     "assets": [
>         "node_modules/@img/sharp-darwin-arm64/**/*",
>         "node_modules/@img/sharp-libvips-darwin-arm64/**/*"
>     ]
> }
> ```

**Comment by LetrixZ** [comment]
> Sorry to revive a closed thread, potential fix described here [#4377 (comment)](https://github.com/lovell/sharp/issues/4377#issuecomment-2819585265)
> 
> > Just a quick update. Found a solution. I was missing `--config package.json` in the `npx` command:
> > ```
> > npx pkg index.js --config package.json --targets node20-macos-arm64 --output index
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And added assets to `package.json`:
> > ```
> >  "pkg": {
> >     "assets": [
> >         "node_modules/@img/sharp-darwin-arm64/**/*",
> >         "node_modules/@img/sharp-libvips-darwin-arm64/**/*"
> >     ]
> > }
> > ```

Does this works if the user doesn't have `libvips` installed in their system?

---

UPDATE: Checked `@yao-pkg/pkg` and it works by bundling the files and extracting them at runtime to `$HOME/.cache` and I could only get it working with the prebuilt binaries and not with custom Vips because it can't load `libvips-cpp.so.42`.

**Comment by Teyik0** [comment]
I switched to @napi-rs/image and got it working with Bun by manually copying the N-API .node binding next to the compiled binary.

Bun doesn’t bundle native bindings, so this build.ts just locates the .node file (even through symlinks in a monorepo) and copies it to dist/.

Might work with sharp too, but I couldn’t get it stable.

```ts
import { realpathSync } from "node:fs";
import { basename, dirname, resolve } from "node:path";

await Bun.build({
	entrypoints: ["src/index.ts"],
	outdir: "dist",
	compile: {
		outfile: "server",
	},
	minify: {
		whitespace: true,
		syntax: true,
	},
	target: "bun",
});

const pattern = new Bun.Glob("**/*.node");
const napiImageRealPath = dirname(
	realpathSync(resolve(import.meta.dir, "node_modules/@napi-rs/image")),
);

const files = Array.from(
	pattern.scanSync({
		cwd: napiImageRealPath,
		followSymlinks: true,
		absolute: true,
	}),
);

const fileName = basename(files[0]);
const destPath = `dist/${fileName}`;

const file = Bun.file(files[0]);
if (!(await file.exists()))
	throw new Error("@napi-rs/image .node napi file binding not found");

await Bun.write(destPath, file);
```

This work with mostly this minimum image, in case you want to dockerize.
**gcr.io/distroless/cc-debian12**
